### PR TITLE
serverexperimentation: new proto related backend changes

### PR DESCRIPTION
**Pull request by kathan24**
### Description

These are the backend serverexperimentation changes related to new proto changes in [this PR](https://github.com/lyft/clutch/pull/699)

### Testing Performed

Tested locally 

### GitHub Issue

Fixes #

### CODE REVIEW

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 1]
```diff
@@ -186,11 +185,27 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
 		for _, experiment := range experiments {
-			testConfig := &serverexperimentation.TestConfig{}
-			if !isFaultTest(experiment, testConfig) {
+			httpFaultConfig := &serverexperimentation.HTTPFaultConfig{}
+			if !isFaultTest(experiment, httpFaultConfig) {
 				continue
 			}
-			percentageKey, percentageValue, faultKey, faultValue := createRuntimeKeys(testConfig, ingressPrefix, egressPrefix, logger)
+
+			upstreamCluster, downstreamCluster, err := unParse(httpFaultConfig)
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:193]: can you think of a better name than `unParse`? `getClusterPair` perhaps?

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 2]
```diff
@@ -186,11 +185,27 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
+
+			var enforcer string
+			switch httpFaultConfig.GetFaultTargeting().GetEnforcer().(type) {
+			case *serverexperimentation.FaultTargeting_DownstreamEnforcing:
+				enforcer = "downstreamEnforcing"
+			case *serverexperimentation.FaultTargeting_UpstreamEnforcing:
+				enforcer = "upstreamEnforcing"
+			default:
+				logger.Errorw("unknown enforcer %v", httpFaultConfig)
+			}
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:206]: maybe move this into a function and call it inline during logging? it kinda disrupts the flow of the function atm

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 3]
```diff
@@ -186,11 +185,27 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
 	// No experiments meaning clear all experiments for the given upstream cluster
 	if len(experiments) != 0 {
 		for _, experiment := range experiments {
-			testConfig := &serverexperimentation.TestConfig{}
-			if !isFaultTest(experiment, testConfig) {
+			httpFaultConfig := &serverexperimentation.HTTPFaultConfig{}
+			if !isFaultTest(experiment, httpFaultConfig) {
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:189]: while you're in here wdyt about changing `isFaultTest` to be something like `maybeUnmarshalFaultTest(experiment) (*serverexperimentation.HttpFaultConfig, bool)`? The function name reads like a predicate but it actually mutates the input params which seems a bit odd

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 4]
```diff
@@ -186,11 +185,27 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
 				continue
 			}
-			percentageKey, percentageValue, faultKey, faultValue := createRuntimeKeys(testConfig, ingressPrefix, egressPrefix, logger)
+
+			upstreamCluster, downstreamCluster, err := unParse(httpFaultConfig)
+			if err != nil {
+				logger.Errorw("Invalid http fault config", "config", httpFaultConfig)
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:195]: This and the other error conditions below should probably `continue` or `return`?

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 5]
```diff
@@ -242,86 +257,83 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
 
 		default:
-			logger.Errorw("Invalid fault injection cluster found", "upstream", target.GetUpstreamCluster(), "downstream", target.GetDownstreamCluster(), "faultInjectionCluster", faultInjectionCluster)
+			logger.Errorw("unknown enforcer %v", httpFaultConfig)
 		}
 
 	default:
-		logger.Errorw("Unknown fault type: %t", testConfig)
+		logger.Errorw("Unknown fault type: %t", httpFaultConfig)
 		panic("Unknown fault type")
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:321]: Out of scope for this PR but seems bad to crash here

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 6]
```diff
@@ -242,86 +257,83 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
-		percentageValue = latency.Percent
-		faultValue = latency.DurationMs
+	case *serverexperimentation.HTTPFaultConfig_LatencyFault:
+		latency := httpFaultConfig.GetLatencyFault()
+		percentageValue = latency.GetPercentage().GetPercentage()
+		faultValue = latency.GetLatencyDuration().GetFixedDurationMs()
 
-		switch faultInjectionCluster {
-		case serverexperimentation.FaultInjectionCluster_FAULTINJECTIONCLUSTER_DOWNSTREAM:
+		switch interface{}(httpFaultConfig.GetFaultTargeting().GetEnforcer()).(type) {
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:298]: is the cast to empty interface necessary here?

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache_test.go [hunk 7]
```diff
@@ -209,57 +237,59 @@ func TestCreateRuntimeKeys(t *testing.T) {
...
-				expectedPercentageKey = fmt.Sprintf(HTTPPercentageForExternal, egressPrefix, target.UpstreamCluster)
-				expectedFaultKey = fmt.Sprintf(HTTPStatusForExternal, egressPrefix, target.UpstreamCluster)
-			} else if faultInjectionType == serverexperimentation.FaultInjectionCluster_FAULTINJECTIONCLUSTER_UPSTREAM {
-				if target.DownstreamCluster == "" {
+		case *serverexperimentation.HTTPFaultConfig_AbortFault:
+			abort := config.GetAbortFault()
+			expectedFaultValue = abort.GetAbortStatus().GetHttpStatusCode()
+			expectedPercentageValue = abort.GetPercentage().GetPercentage()
+
+			switch interface{}(config.GetFaultTargeting().GetEnforcer()).(type) {
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache_test.go:259]: It might be nice to add some extra helper functions here and elsewhere just to reduce some of the nesting in this function, it's a bit hard to follow

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache.go [hunk 8]
```diff
@@ -242,86 +257,83 @@ func setSnapshot(snapshotCache cacheWrapper, rtdsLayerName string, cluster strin
...
 				// Latency Internal Fault for a given downstream services
-				percentageKey = fmt.Sprintf(LatencyPercentageWithDownstream, ingressPrefix, target.GetDownstreamCluster())
-				faultKey = fmt.Sprintf(LatencyDurationWithDownstream, ingressPrefix, target.GetDownstreamCluster())
+				percentageKey = fmt.Sprintf(LatencyPercentageWithDownstream, ingressPrefix, downstreamCluster)
+				faultKey = fmt.Sprintf(LatencyDurationWithDownstream, ingressPrefix, downstreamCluster)
 			}
 
 		default:
-			logger.Errorw("Invalid fault injection cluster found", "upstream", target.GetUpstreamCluster(), "downstream", target.GetDownstreamCluster(), "faultInjectionCluster", faultInjectionCluster)
+			logger.Errorw("unknown enforcer %v", httpFaultConfig)
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/rtds/rtds_cache.go:316]: wdyt about returning errors on these error conditions so we bail out on generating runtime keys in this case?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks like a good start! 

There seems to be a lot of error conditions, are there tests covering that?

File: backend/module/chaos/serverexperimentation/rtds/rtds_cache_test.go (hunk 7 above)
Comment (kathan24) [reply backend/module/chaos/serverexperimentation/rtds/rtds_cache_test.go]: This nesting has different logics within it. I could make a func for this specific case which can be used in `rtds_cahce.go` but I don't see much benefit from it. Let me know your thought

File: backend/module/chaos/serverexperimentation/server_experimentation.go [hunk 9]
```diff
@@ -57,8 +57,47 @@ func (s *Service) transform(_ *experimentstore.ExperimentRun, config *experiment
 		return nil, err
 	}
 
-	upstream := experimentConfig.GetClusterPair().GetUpstreamCluster()
-	downstream := experimentConfig.GetClusterPair().GetDownstreamCluster()
+	var downstream, upstream string
+	switch interface{}(experimentConfig.GetFaultTargeting().GetEnforcer()).(type) {
```
Comment (snowp) [inline backend/module/chaos/serverexperimentation/server_experimentation.go:61]: Remove the interface{} cast here and below

File: frontend/workflows/experimentation/src/tests/list-experiments.test.tsx [hunk 10]
```diff
@@ -11,16 +11,17 @@ jest.mock("react-router-dom", () => {
 });
```
Comment (snowp) [inline frontend/workflows/experimentation/src/tests/list-experiments.test.tsx:11]: Are the frontend changes here intended?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks for iterating, this is looking good. Are the frontend changes in this PR intentional? Should update the PR title if so

**Review by snowp — APPROVED** [review]
